CHANGELOG.md

```diff
@@ -50,6 +51,7 @@
 - [`pkg/test.FrameworkClient`](https://github.com/operator-framework/operator-sdk/blob/master/pkg/test/client.go#L33) methods `List()` and `Delete()` have new signatures corresponding to the homonymous methods of `sigs.k8s.io/controller-runtime/pkg/client.Client`. ([#1876](https://github.com/operator-framework/operator-sdk/pull/1876))
 - CRD file names were previously of the form `<group>_<version>_<kind>_crd.yaml`. Now that CRD manifest `spec.version` is deprecated in favor of `spec.versions`, i.e. multiple versions can be specified in one CRD, CRD file names have the form `<full group>_<resource>_crd.yaml`. `<full group>` is the full group name of your CRD while `<group>` is the last subdomain of `<full group>`, ex. `foo.bar.com` vs `foo`. `<resource>` is the plural lower-case CRD Kind found at `spec.names.plural`. ([#1876](https://github.com/operator-framework/operator-sdk/pull/1876))
 - Upgrade Python version from `2.7` to `3.6`, Ansible version from `2.8.0` to `~=2.8` and ansible-runner from `1.2` to `1.3.4` in the Ansible based images. ([#1947](https://github.com/operator-framework/operator-sdk/pull/1947)) 
+- Removed `kubemetrics.NewCollectors` and changed exported function signature for `kubemetrics.ServeMetrics` due to a [breaking change in kube-state-metrics](https://github.com/kubernetes/kube-state-metrics/pull/786). ([#1943](https://github.com/operator-framework/operator-sdk/pull/1943))  
```

Comment:
As we change `kubemetrics.NewCollectors()` to `kubemetrics.NewMetricsStores()` I'm wondering why this is an API exposed to users in the first place.

Right now the `cmd/manager/main.go` scaffold is just expected to call `kubemetrics.GenerateAndServeCRMetrics()` for recording CR counts.

If users are expected to make use of `kubemetrics.NewMetricsStores()` for recording custom metrics then there should be an example or documentation for that. Otherwise it doesn't make sense to keep this public.

Also we can address this as a follow up but thought I'd highlight this issue here since it seems like changing `kubemetrics.NewCollectors` isn't really a breaking change unless we've documented this as a public API.

Reply:
Also can we change `kubemetrics.NewCollectors()` to `pkg/kube-metrics.NewCollectors()` since that's the actual pkg. And also if we're expecting users to use `NewCollectors()` then we should point out what they should change it to `pkg/kube-metrics.NewMetricsStores()`.

Reply:
There was [a lot of discussion in the original PR](https://github.com/operator-framework/operator-sdk/pull/1277) about what to export (or not). I think the thought at the time was to export `kubemetrics.GenerateAndServeCRMetrics()` for use in `main.go` to simplify things for users by default, but to also export `kubemetrics.NewCollectors()` and other functions so that advanced users could have more fine-grained control.

In general I would typically lean toward exporting a smaller API surface and opening things up more later, if asked. For the `kube-metrics` package in particular, I think the API surface could use some improvements and refactoring to make it simple for base cases but more flexible for advanced cases. Perhaps we do that in a follow up or if/when we upstream it?

With that in mind, I think we should leave things as untouched as possible. I'll update this line to say that `NewCollectors` has been replaced by `NewMetricsStores`. 

WDYT?

Reply:
> it seems like changing kubemetrics.NewCollectors isn't really a breaking change unless we've documented this as a public API.

Since users could theoretically be importing and using exported functions, I would generally consider any breaking change to any exported API to be a breaking change even if we don't use that API directly in a scaffold or document that it's public.

Reply:
Agreed. We can refine/document this better later.